### ISSUE TITLE
Re-structure the build so we do not need source inclusion anymore

### ISSUE THREAD

**Issue by gzm0** [scala-js/scala-js#3714]
Using the techiniques in https://github.com/gzm0/sbtception we can use things without defining their build twice.

This is an alternative to #3661.

**Comment by gzm0** [comment]
I wanted to have a `build` directory and hide most of the things in there. Turns out sbt is quite strict with project directories being sub directories of the root. In practice, this means we'll need to actually move code (unless we want to go even more non-standard).

I see two main options.

### Minimal

Basically a build `core` containing everything up-to and including `sbt-plugin` and a build `full` containing the rest (tests, testing infra, cross builds of stuff in core).

### Semantic

Use this opportunity to group stuff. Rough idea:

- `js-env`: all jsenv stuff including test kit (could also be `core`).
- `core`: compiler plugin, linker, sbt-plugin, libraries, test bridge JVM
- `testing`: jUnit plugin, test bridge JS.
- `tests`: partest, test suite (ex), test suite linker
- `examples`: examples (already here).
- `core-js`: JS build of core stuff.

Both of these options will need an additional `build-common` build containing stuff needed across all builds.

I have a slight preference for the second one, since I believe it will make navigating the SJS source tree easier (not sure compared to now, but definitely compared to `core` and `full` only).

@sjrd WDYT?

**Comment by sjrd** [comment]
> I wanted to have a `build` directory and hide most of the things in there. Turns out sbt is quite strict with project directories being sub directories of the root. In practice, this means we'll need to actually move code (unless we want to go even more non-standard).

Could we not simply have the `project`s themselves be subdirectories of the root, but require their `unmanagedSourceDirectories` to point outside? It seems like this would have minimal changes. We would need something like that for `core-js` anyway, wouldn't we?

**Comment by sjrd** [comment]
Regardless, I would suggest that we keep the number of *builds* to the minimum. So I would advocate for the **Minimal** solution above.

Additionally, I would put `libraries` in the "full" build rather than the "core". Except of course the project that contains only `RuntimeLong` (once it arrives).

**Comment by gzm0** [comment]
> Could we not simply have the projects themselves be subdirectories of the root, but require their unmanagedSourceDirectories to point outside? It seems like this would have minimal changes.

That would mean a non-standard layout and more changes, but yes, that would probably be doable.

> Regardless, I would suggest that we keep the number of builds to the minimum.

Is there any particular reason for that?

> Additionally, I would put libraries in the "full" build rather than the "core". Except of course the project that contains only RuntimeLong (once it arrives).

We can't really do that, because the unit tests of the linker depend on the libraries.

**Comment by sjrd** [comment]
>  Is there any particular reason for that?

Sanity? Also it's unclear that there exists a semantic decomposition that is objectively better than others. So I believe the object decomposition of "as little as possible in the `core`, everything else in `full` (or root)" is a good, enforceable criteria.

> We can't really do that, because the unit tests of the linker depend on the libraries.

Ah ... That's a bummer. OK then.

**Comment by gzm0** [comment]
An update on this: I have been trying to get something like https://github.com/gzm0/sbtception to work in the Scala.js build, but I'm getting stuck everywhere. At a sufficient level of nesting, things just stop compiling and the right things are not on the classpath anymore.

So I am not sure it will be possible to pull this off at all. I will try again with sbt 1.2.x, probably over the rainy weekend, but we might have to start investigating other options.

There are two further options I see:
- Use scripted for scripted for all tests (IIRC, we only use the linker in tests).
- Load `linker.<package>` and `linker.StandardLinker` via reflection, but not `linker._`: This would allow us to keep the same interface in the sbt plugin.

**Comment by sjrd** [comment]
I don't understand your second alternative. How could we load by reflection part of the linker but not the rest? And how would that help?

**Comment by gzm0** [comment]
> How could we load by reflection part of the linker but not the rest? And how would that help?

We could compile statically against the *interface* but load the *implementation* via reflection. It would help because:

- The types in the sbt plugin would still be nice / the same.
- We can load the implementation (and hence the resources) in a private class loader.
- Users of the sbt plugin can build against the normal linker and replace parts if necessary without even knowing that it is reflectively loaded in the plugin.

How to do this would be something like:
- Manually add source files for the linker interface (`Linker`, `LinkerConfig`) to the sbt plugin compile.
- Create a weirdo class loader that loads interface classes from the parent, but the implementations from a supplied JAR. Use this class loader to create a value for `scalaJSLinker`.

Alternatively we could even have two artifacts: `scalajs-linker-interface` and `scalajs-linker`.

**Comment by sjrd** [comment]
Ah I see now. OK, that makes sense. To make it clean, I believe we should indeed have two artifacts in that case. Otherwise it will be too easy for our sbt plugin or a dependent sbt plugin to shoot themselves in the foot: they would try and use stuff from the "implementation" portions (including stuff in `standard._`), but that wouldn't be the same classes than those loaded by reflection. If however they can compile against `scalajs-linker-interface`, then the code is guaranteed to work.

**Comment by sjrd** [comment]
Hum ... actually I don't see how this will work in practice. We would need to completely dissociate the interface from the implementation, otherwise the interface could have `LinkageError`s because it cannot resolve stuff from the implementation in the same classloader.

For example, `StandardLinker.apply()`, which is in `linker._`, calls into `standard.StandardLinkerImpl`. If only the interface is in the plugin classloader, calling `StandardLinker.apply()` will statically compile, but will cause a `LinkageError` at run-time.

Similarly, calling `IRFile.fromPath()`, which is available in `linker._`, would transitively depend on `standard.IRFileImpl` and also cause a `LinkageError`.

Currently our entire API is built on a model where the *API* of super-stable stuff is independent from the stable or private stuff; but its implementation is dependent.

**Comment by gzm0** [comment]
> Currently our entire API is built on a model where the API of super-stable stuff is independent from the stable or private stuff; but its implementation is dependent.

Yes. As I look at this, this is what I realize as well: We'd have to even further separate the API.

- Interface.
- Super-stable implementation.
- Less stable stuff.

The last two bullets can (and probably should) be in the same artifact. Alternatively, we do this separation ad-hoc (probably with name filtering), but that is indeed not very nice.

**Comment by gzm0** [comment]
I think I managed to come up with a separation that works. WIP at https://github.com/gzm0/scala-js/tree/linker-interface

However, there is another problem: Most of the things we want to load reflectively are used in settings. However, in our own build, they need to depend on tasks of other projects (to get the actual linker JAR). Now, the main reason they are settings is because we want them to be persistent across runs, which is arguably a hack.

I have tried to find ways to cache things, notably on sbt's `State`. But I couldn't find a way to make it work (n.b. I couldn't find a pre-task execution command hook). @sjrd did you have any ideas how to do this, when you started the reflective interface to the linker?

**Comment by sjrd** [comment]
> did you have any ideas how to do this, when you started the reflective interface to the linker?

Perhaps look at this comment from Lukas: https://github.com/sjrd/sbt-dynscalajs/issues/7#issuecomment-332075451. He commented on sbt-dynscalajs which has the same problem.

I also had the idea of using a setting of type `class Box[A](var value: A)`, and have a task that uses that box as a cache, and otherwise computes something and sets the value in the box for next time. Not pretty, but it should work relatively easily.

**Comment by gzm0** [comment]
Well, that sbt code seems to hook directly into sbt's loading mechanism... We do not have that luxury.

As for using a box: I cannot come up with a solution that needs less than three keys if we want to allow both usage and changing the linker easily. Basically we'd need:

- Setting with thunk to create a linker.
- Magic setting that caches stuff.
- Task that actually accesses the linker and passes in relevant stuff.

Now in its simplest form, this will even leak implementation details of what we need to create the linker (because we pass it as parameters from the last task). This might be fixable by using sbt foo.

Still, it is kind of unsatisfactory. Any better ideas?

**Comment by gzm0** [comment]
Closing in favor of #3661.
